### ISSUE TITLE
*nix OuterLoops involving certificates fail to find the CRL during certificate validation

### ISSUE THREAD

**Issue by roncain** [dotnet/wcf#1174]
When running OuterLoop tests on an Ubuntu machine that has been manually setup (against a self-hosted WCF service running on a Windows machine), it appears the certificate ConditionalFact detector/installer has successfully installed the certificates.  But they are considered invalid because the CRL itself cannot be reached during certificate validation.

Possible reasons are a wrong URL for the CRL service inside the certificate or a failure to respond properly.

```
         Discovered:  Binding.Custom.Tests
         Starting:    Binding.Custom.Tests
            CustomBindingTests.DefaultSettings_Tcp_Binary_Echo_RoundTrips_String [FAIL]
               System.ServiceModel.Security.SecurityNegotiationException : The X.509 certificate CN=roncain7a.redmond.corp.microsoft.com, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 2F2AB3AE9BB90BF742049481D99A7A35FAE4A9B5 chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL 
               ---- System.IdentityModel.Tokens.SecurityTokenValidationException : The X.509 certificate CN=roncain7a.redmond.corp.microsoft.com, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 2F2AB3AE9BB90BF742049481D99A7A35FAE4A9B5 chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL 

```

**Comment by zhenlan** [comment]
@iamjasonp can you please try to run these tests on *nix at least manually before RTM?

**Comment by iamjasonp** [comment]
I checked this manually but haven't run the tests yet. On first glance, I don't see why this is failing, and I will run the tests manually to check out what's happening.

**Comment by roncain** [comment]
The test shown in the description passed for me today running manually on Ubuntu (5/31), but I still see the same failure in Helix runs:
```
        System.ServiceModel.Security.SecurityNegotiationException : The X.509 certificate CN=WCFCORESRV4.cloudapp.net, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 5BE6E3C70A2AFA62FA63A745099020E89D0FF6FF chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL \n---- System.IdentityModel.Tokens.SecurityTokenValidationException : The X.509 certificate CN=WCFCORESRV4.cloudapp.net, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 5BE6E3C70A2AFA62FA63A745099020E89D0FF6FF chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL 
```

**Comment by roncain** [comment]
I see these failures in every nightly Helix run of *nix flavors.  Incidentally, the affected tests are:
```
Binding.Custom.Tests#CustomBindingTests.DefaultSettings_Tcp_Binary_Echo_RoundTrips_String 
Security.TransportSecurity.Tests#IdentityTests.ServiceIdentityNotMatch_Throw_MessageSecurityException 
Security.TransportSecurity.Tests#IdentityTests.VerifyServiceIdentityMatchDnsEndpointIdentity 
Security.TransportSecurity.Tests#Tcp_ClientCredentialTypeCertificateCanonicalNameTests.TcpClientCredentialType_Certificate_With_Only_CanonicalName_EchoString(endpointUri: net.tcp://wcfcoresrv4.cloudapp.net:809//TcpCertificate... 
Security.TransportSecurity.Tests#Tcp_ClientCredentialTypeCertificateCanonicalNameTests.TcpClientCredentialType_Certificate_With_Only_CanonicalName_EchoString(endpointUri: net.tcp://wcfcoresrv4.cloudapp.net:809//TcpCertificate...-0 
Security.TransportSecurity.Tests#Tcp_ClientCredentialTypeCertificateCanonicalNameTests.TcpClientCredentialType_Certificate_With_Only_CanonicalName_EchoString(endpointUri: net.tcp://wcfcoresrv4.cloudapp.net:809//TcpCertificate...-1 
Security.TransportSecurity.Tests#Tcp_ClientCredentialTypeTests.TcpClientCredentialType_Certificate_With_ServerAltName_EchoString 

```

**Comment by iamjasonp** [comment]
Very interesting - this is not reproing for me when I'm trying these out.

**Comment by roncain** [comment]
@iamjasonp -- how are you starting the self-hosted service when you test?  Using PsExec?  And are you following the manual setup guide?  I still have issues following the setup steps as documented

**Comment by iamjasonp** [comment]
> @iamjasonp -- how are you starting the self-hosted service when you test? Using PsExec? And are you following the manual setup guide? I still have issues following the setup steps as documented

I've tested starting the service as both LOCAL SYSTEM (psexec) and as a normal user credential. I think the setup steps are now a bit out of date; let's take the effort to update them. 

All that said, I've tested various permutations of the service/client and am still not able to repro this.

I think that the next course of action is to isolate the certificate tests so that the CRL is hit _only_ by those tests requiring them (that is, only when certs are expected to be revoked). We should also add a positive test for this case.

**Comment by roncain** [comment]
Update: CustomBindingTests/DefaultSettings_Tcp_Binary_Echo_RoundTrips_String fails on OS X in Helix with the identical exception I see in Ubuntu:
```
System.ServiceModel.Security.SecurityNegotiationException : The X.509 certificate CN=WCFCORESRV4.cloudapp.net, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 11045E59349D9799852F564EDBEE4AFF1A457A32 chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL 
---- System.IdentityModel.Tokens.SecurityTokenValidationException : The X.509 certificate CN=WCFCORESRV4.cloudapp.net, O=DO_NOT_TRUST, OU=Created by https://github.com/dotnet/wcf; 11045E59349D9799852F564EDBEE4AFF1A457A32 chain building failed. The certificate that was used has a trust chain that cannot be verified. Replace the certificate or change the certificateValidationMode. unable to get certificate CRL 
```

**Comment by iamjasonp** [comment]
Fixed by #1390 - we now generate certificates with a timestamp in the issuer name. This helps change the hash enough that OpenSSL won't complain about this anymore.
